src/Strategy/Node/Npm/PackageLockV3.hs

```diff
@@ -0,0 +1,435 @@
+              { dependencyType = NodeJSType
+              , dependencyName = unPackageName packageName
+              , dependencyVersion = CEq <$> (plV3PkgVersion meta)
+              , dependencyLocations = catMaybes [unNpmLockV3Resolved $ plV3PkgResolved meta]
+              , dependencyEnvironments =
+                  if plV3PkgDev meta
+                    then Set.singleton EnvDevelopment
+                    else Set.singleton EnvProduction
+              , dependencyTags = Map.empty
+              }
+        _ -> Nothing -- don't report workspace or root project as dependency, or pkg not found
+    allDependencyPackages :: Map PackagePath PackageLockV3Package
+    allDependencyPackages =
+      Map.filterWithKey
+        (\k _ -> isReportableDependencyModule k)
+        (packages pkgLockV3)
+
+    -- True if package path leads to dependency, Otherwise false.
+    isReportableDependencyModule :: PackagePath -> Bool
+    isReportableDependencyModule pkgPath = case pkgPath of
+      PackageLockV3PathKey _ _ -> True
+      _ -> False
+
+    -- Provides candidate path in top level tree
+    -- >> toTopLevelPackage pkgName = PackagePath "node_modules/" pkgName
+    toTopLevelPackage :: PackageName -> PackagePath
+    toTopLevelPackage = PackageLockV3PathKey defaultVendorDir
+
+    -- Provides candidate path in as if the package was vendored
+    -- >> toVendoredPackage "node_module/a" pkgName = PackagePath "node_modules/a/node_modules/" pkgName
+    toVendoredPackage :: Text -> PackageName -> PackagePath
+    toVendoredPackage prefix = PackageLockV3PathKey (prefix <> "/" <> defaultVendorDir)
+
+    -- Provides candidate path in current workspace root, if it contains a workspace root.
+    -- If prefix does not contain workspace root, it mimics @toVendoredPackage.
+    --
+    -- >> toVendoredPackage "myWorkspace/myPkg/node_modules/a" pkgName = PackagePath "myWorkspace/myPkg/node_modules/" pkgName
+    -- >> toVendoredPackage "myWorkspace/myPkg/node_modules/a/node_modules/b" pkgName = PackagePath "myWorkspace/myPkg/node_modules/" pkgName
+    -- >> toVendoredPackage "node_modules/a" pkgName = PackagePath "node_modules/a/node_modules/" pkgName
+    toWorkspaceLevelPackage :: Text -> PackageName -> PackagePath
+    toWorkspaceLevelPackage prefix = PackageLockV3PathKey (rootWorkspace <> "/" <> defaultVendorDir)
+      where
+        rootWorkspace :: Text
+        rootWorkspace = fromMaybe prefix (getRootWorkspace prefix)
+
+        getRootWorkspace :: Text -> Maybe Text
+        getRootWorkspace pkgPath = case Text.breakOn "/node_modules" pkgPath of
+          ("", "") -> Nothing
+          (firstModule, _)
+            | firstModule == pkgPath -> Nothing
+            | (Map.size (Map.filterWithKey (\k _ -> isWorkSpace firstModule k) (packages pkgLockV3)) > 0) -> Just firstModule
+            | otherwise -> Nothing
+
+        isWorkSpace :: Text -> PackagePath -> Bool
+        isWorkSpace name pkgPath = pkgPath == PackageLockV3WorkSpace name
```

Comment:
[nit]: We should generally avoid nested `where` clauses.  The `rootWorkspace` above is fine, but these don't depend on the `prefix` binding, so we can push them out to the top `where` clause to avoid cluttering up the nested one.
```suggestion
    getRootWorkspace :: Text -> Maybe Text
    getRootWorkspace pkgPath = case Text.breakOn "/node_modules" pkgPath of
      ("", "") -> Nothing
      (firstModule, _)
        | firstModule == pkgPath -> Nothing
        | (Map.size (Map.filterWithKey (\k _ -> isWorkSpace firstModule k) (packages pkgLockV3)) > 0) -> Just firstModule
        | otherwise -> Nothing

    isWorkSpace :: Text -> PackagePath -> Bool
    isWorkSpace name pkgPath = pkgPath == PackageLockV3WorkSpace name
```